### PR TITLE
chore: add azure sp login

### PR DESCRIPTION
**Pull request by xiaolang124**
1. Add service principal login for CLI Azure
2. Save encrypted service principal data in ~/home/account/azure.sp

image

### CODE REVIEW

File: packages/cli/src/cmds/account.ts [hunk 1]
```diff
@@ -169,14 +181,37 @@ class AccountLogin extends YargsCommand {
         description: "Authenticate with a specific Azure Active Directory tenant.",
         type: "string",
         default: "",
+      })
+      .options("service-principal", {
```
Comment (swatDong) [inline packages/cli/src/cmds/account.ts:185]: Is this to follow the same behavior as az-cli? Add error handlings for all unsupported cases:
- -u xxx@xxx.com -p SECRET (user login)
- --service-principal -u CLIENTID -p CERTIFICATE -t TENANTID (cert credential)
- --service-principal -u CLIENTID -p SECRET (no tenant)

Call out in documentation/example that the only supported case is **--service-principal -u CLIENTID -p SECRET -t TENANTID**

Comment (swatDong) [inline packages/cli/src/cmds/account.ts:185]: Anyway to show these options only for **azure**?

File: packages/cli/src/cmds/account.ts [hunk 2]
```diff
@@ -169,14 +181,37 @@ class AccountLogin extends YargsCommand {
...
+      })
+      .options("username", {
+        alias: "u",
+        description: "Client ID for service principal",
+        type: "string",
+        default: "",
+      })
+      .options("password", {
+        alias: "p",
+        description: "Client ID for service principal",
```
Comment (swatDong) [inline packages/cli/src/cmds/account.ts:198]: Any plan to support certificate credential? Certificate is more commonly used than secret in production environments. And to meet security requirements as well.

File: packages/cli/src/cmds/account.ts (hunk 1 above)
Comment (xiaolang124) [reply packages/cli/src/cmds/account.ts:217]: Seperate login command to m365 and azure.

Comment (xiaolang124) [reply packages/cli/src/cmds/account.ts:223]: Already add parameter check logic. And add examples to command help.
image

File: packages/cli/src/cmds/account.ts (hunk 2 above)
Comment (xiaolang124) [reply packages/cli/src/cmds/account.ts]: Add cert login logic to CLI.

File: packages/cli/src/commonlib/azureLoginCI.ts [hunk 3]
```diff
@@ -98,53 +154,155 @@ export class AzureAccountManager extends login implements AzureAccountProvider {
...
     throw new Error("Method not implemented.");
   }
 
-  getJsonObject(showDialog?: boolean): Promise<Record<string, unknown> | undefined> {
-    throw new Error("Method not implemented.");
+  async listSubscriptions(): Promise<SubscriptionInfo[]> {
+    const credential = await this.getAccountCredentialAsync();
+    if (credential) {
+      const client = new SubscriptionClient(credential);
+      const tenants = await listAll(client.tenants, client.tenants.list());
```
Comment (swatDong) [inline packages/cli/src/commonlib/azureLoginCI.ts:181]: SP is bound in one tenant, why need to list tenants for SP?

File: packages/cli/src/commonlib/cacheAccess.ts [hunk 4]
```diff
@@ -201,3 +202,54 @@ export async function loadAccountId(accountName: string) {
...
   return undefined;
 }
+
+export async function saveAzureSP(
+  clientId: string,
+  secret: string,
+  tenantId: string
+): Promise<void> {
+  await fs.ensureDir(cacheDir);
+  const azureSPCrypto = new AccountCrypto("azure-sp");
```
Comment (swatDong) [inline packages/cli/src/commonlib/cacheAccess.ts:212]: Use `this.accountCrypto`, do not create another AccountCrypto.

File: packages/cli/src/commonlib/azureLoginCI.ts (hunk 3 above)
Comment (xiaolang124) [reply packages/cli/src/commonlib/azureLoginCI.ts]: Good point, I have removed multi tenants part.

File: packages/cli/src/commonlib/cacheAccess.ts (hunk 4 above)
Comment (xiaolang124) [reply packages/cli/src/commonlib/cacheAccess.ts]: Thanks for the advice, I have changed ```AzureSpCypto``` method to static method.

File: packages/cli/src/commonlib/cacheAccess.ts [hunk 5]
```diff
@@ -201,3 +202,60 @@ export async function loadAccountId(accountName: string) {
 
   return undefined;
 }
+
+export class AzureSpCrypto {
+  private static readonly accountCrypto: AccountCrypto = new AccountCrypto("sp");
```
Comment (swatDong) [inline packages/cli/src/commonlib/cacheAccess.ts:207]: Just use "azure" as account name to keep consistency with user azure account. No need to add another crypto key for SP.

**Review by swatDong — DISMISSED** [review]

**Review by swatDong — DISMISSED** [review]

**Review by swatDong — APPROVED** [review]

**Review by Alive-Fish — APPROVED** [review]

**Pull request merged**